firedrake/slate/slate.py

```diff
@@ -0,0 +1,538 @@
+            par = lambda x: x
+        else:
+            par = lambda x: "(%s)" % x
+
+        return par("-%s" % self.tensor.__str__(prec=self.prec))
+
+
+class BinaryOp(TensorBase):
+    """An abstract SLATE class representing binary operations on tensors.
+    Such operations take two operands and returns a tensor-valued expression.
+
+    The currently supported binary operations include the following:
+
+        (1) The addition of two identical-rank :class:`TensorBase` objects. That is,
+            the addition of two scalars, vectors or matrices. This operation is
+            implemented in the subclass `TensorAdd`.
+        (2) The subtraction of two scalar, vector or matrix expressions. See `TensorSub`.
+        (3) The multiplication of two matrices, the action of a matrix on a vector or
+            the action of a scalar on any other rank tensor. All cases are handled by the
+            `TensorMul` subclass.
+
+    :arg A: a :class:`TensorBase` object. This can be a terminal tensor object
+            (:class:`Tensor`) or any derived expression resulting from any number
+            of linear algebra operations on `Tensor` objects. For example,
+            another instance of a `BinaryOp` object is an acceptable input, or
+            a `UnaryOp` object.
+    :arg B: a :class:`TensorBase` object.
+    """
+
+    def __init__(self, A, B):
+        """Constructor for the BinaryOp class."""
+        super(BinaryOp, self).__init__()
+        self.tensors = A, B
+
+    def coefficients(self):
+        """Returns the expected coefficients of the resulting tensor
+        of performing a binary operation on two tensors. Note that
+        the coefficients are handled the same way for all binary operations.
+        """
+        A, B = self.tensors
+        clist = []
+        A_coeffs = A.coefficients()
+        uniqueAcoeffs = set(A_coeffs)
+        for c in B.coefficients():
+            if c not in uniqueAcoeffs:
+                clist.append(c)
+
+        return tuple(list(A_coeffs) + clist)
+
+    def ufl_domains(self):
+        """Returns the integration domains of the integrals associated with
+        the tensor.
+        """
+        A, B = self.tensors
+        return join_domains(A.ufl_domains() + B.ufl_domains())
+
+    def subdomain_data(self):
+        """Returns a mapping on the tensor:
+        `{domain:{integral_type: subdomain_data}}`.
+        """
+        A, B = self.tensors
+        if A.rank == 0:
+            return B.subdomain_data()
+        elif B.rank == 0:
+            return A.subdomain_data()
+        else:
+            # Join subdomain_data
+            sd_dict = A.subdomain_data()[A.ufl_domain()]
+            for sd_key_B in B.subdomain_data()[B.ufl_domain()].keys():
+                if sd_key_B not in sd_dict.keys():
+                    sd_dict.update({sd_key_B: B.subdomain_data()[B.ufl_domain()][sd_key_B]})
+
+            return {self.ufl_domain(): sd_dict}
```

Comment:
No. For example, when constructing the RHS of a hybridized problem, you need to take the matrix associated with the lagrange multipliers (a `Tensor` constructed from a `dS` form) and multiply against a vector associated with a `dx` form. In this case, if I `assert A.subdomain_data() == B.subdomain_data()`, it will fail due to the difference in `integral_type`.